packages/webview_flutter/webview_flutter_android/lib/src/v4/src/android_webview_controller.dart

```diff
@@ -0,0 +1,385 @@
+@immutable
+class AndroidJavaScriptChannelParams extends JavaScriptChannelParams {
+  /// Constructs a [AndroidJavaScriptChannelParams].
+  AndroidJavaScriptChannelParams({
+    required super.name,
+    required super.onMessageReceived,
+    @visibleForTesting
+        AndroidWebViewProxy webViewProxy = const AndroidWebViewProxy(),
+  })  : assert(name.isNotEmpty),
+        _javaScriptChannel = webViewProxy.createJavaScriptChannel(
+          name,
+          postMessage: withWeakRefenceTo(
+            onMessageReceived,
+            (WeakReference<void Function(JavaScriptMessage)> weakReference) {
+              return (
+                String message,
+              ) {
+                if (weakReference.target != null) {
+                  weakReference.target!(
+                    JavaScriptMessage(message: message),
+                  );
+                }
+              };
+            },
+          ),
+        );
+
+  /// Constructs a [AndroidJavaScriptChannelParams] using a
+  /// [JavaScriptChannelParams].
+  AndroidJavaScriptChannelParams.fromJavaScriptChannelParams(
+    JavaScriptChannelParams params, {
+    @visibleForTesting
+        AndroidWebViewProxy webViewProxy = const AndroidWebViewProxy(),
+  }) : this(
+          name: params.name,
+          onMessageReceived: params.onMessageReceived,
+          webViewProxy: webViewProxy,
+        );
+
+  final android_webview.JavaScriptChannel _javaScriptChannel;
+}
+
+/// Object specifying creation parameters for creating a [AndroidWebViewWidget].
+///
+/// When adding additional fields make sure they can be null or have a default
+/// value to avoid breaking changes. See [PlatformWebViewWidgetCreationParams] for
+/// more information.
+@immutable
+class AndroidWebViewWidgetCreationParams
+    extends PlatformWebViewWidgetCreationParams {
+  /// Creates [AndroidWebWidgetCreationParams].
+  AndroidWebViewWidgetCreationParams({
+    super.key,
+    required super.controller,
+    super.layoutDirection,
+    super.gestureRecognizers,
+    @visibleForTesting InstanceManager? instanceManager,
+  }) : _instanceManager = instanceManager ?? JavaObject.globalInstanceManager;
+
+  /// Constructs a [WebKitWebViewWidgetCreationParams] using a
+  /// [PlatformWebViewWidgetCreationParams].
+  AndroidWebViewWidgetCreationParams.fromPlatformWebViewWidgetCreationParams(
+    PlatformWebViewWidgetCreationParams params, {
+    InstanceManager? instanceManager,
+  }) : this(
+          key: params.key,
+          controller: params.controller,
+          layoutDirection: params.layoutDirection,
+          gestureRecognizers: params.gestureRecognizers,
+          instanceManager: instanceManager,
+        );
+
+  // Maintains instances used to communicate with the native objects they
+  // represent.
+  final InstanceManager _instanceManager;
+}
+
+/// An implementation of [PlatformWebViewWidget] with the Android WebView API.
+class AndroidWebViewWidget extends PlatformWebViewWidget {
```

Comment:
Did you mean to add the implementation of `PlatformWebViewWidget` to this PR? I thought you separated it into another one. It's fine to add it to this one, I just didn't see any tests for it.

Reply:
No, I didn't mean to do that. Sorry about that, removed it.